### PR TITLE
Deprecate @polyconstraint in favor of @constraint

### PR DESCRIPTION
**Pull request by blegat**
We now need to write
```julia
@constraint(m, p >= 0, domain=@set x^2 + y^2 == 1 && x + y >= 0)
```
instead of
```julia
@polyconstraint(m, p >= 0, domain=x^2 + y^2 == 1 && x + y >= 0)
```

### CODE REVIEW

**Comment by joehuchette** [comment]
I'll try to review this tomorrow (at a conference), but I'm not sure how I feel about nesting the ``@constraint`` and ``set`` macros inside each other.

File: src/macros.jl [hunk 1]
```diff
@@ -2,7 +2,7 @@ using JuMP
 import JuMP: getvalue, validmodel, addtoexpr_reorder
 using Base.Meta
 
-export @polyvariable, @polyconstraint
+export Poly, @set, @polyvariable
```
Comment (blegat) [inline src/macros.jl]: `Poly` is just an object used to dispatch `constructvariable!` to PolyJuMP

**Comment by blegat** [comment]
I like the handler idea. However, the `@set` may still be helpful since it permits to create an algebraic set outside the `@constraint` macro (e.g. if you want to create 2 constraints with the same domain). We could maybe extend the syntax like so:
```
<valid_expr> ::=
      <valid_expr> && <valid_expr>
    | <expr1> <= <expr2> -> creates inequality with <expr2> - <expr1>
    | <expr1> >= <expr2> -> creates inequality with <expr1> - <expr2>
    | <expr1> == <expr2> -> creates equality with <expr1> - <expr2>
    | <algset> -> where set is an algebraic set of a basic semialgebraic set
```
That way we could do
```julia
dom = @set x^2 + y^2 == 1 && x + y <= 1
@constraint ... domain = dom
@constraint ... domain = dom && x >= y
@constraint ... domain = x + 1 && x^2 + y
```
The `@set` constraint could also be moved to [MultivariatePolynomials](https://github.com/blegat/MultivariatePolynomials.jl) at some point.

File: src/macros.jl (hunk 1 above)
Comment (blegat) [inline src/macros.jl]: Oh I get it now :D
It has nothing to do in this PR, it must have messed up my rebase ^^

**Comment by joehuchette** [comment]
I agree that the ``@set`` macro is independently useful. The handler in JuMP could essentially just call out to ``@set``.

@mlubin, thoughts on adding a Dict in JuMP's macro code that contains functions to be called at macro expansion time for unrecognized keyword arguments?

**Comment by mlubin** [comment]
It seems a bit problematic for the expansion of the macro to depend on which packages besides JuMP are loaded at compile time. Also, what if two JuMP extensions want to do something different with the ``domain`` argument?

**Comment by joehuchette** [comment]
> It seems a bit problematic for the expansion of the macro to depend on which packages besides JuMP are loaded at compile time. 

Conceptually it's not ideal, but you would need the package loaded to do anything meaningful to begin with. The only case where I can really see this being an issue is if...

> two JuMP extensions want to do something different with the domain argument?

A nice error, hopefully. Though this would probably be an indication that the two packages aren't compatible to begin with.

One solution would be to dispatch on the ``extra`` pieces as with @blegat recent JuMP PR for variable constructors. So something like:

```
@constraint(m, p >= 0, Poly(), domain= x^2 + y^2 == 1 && x + y >= 0)
```
Cons:
* More verbose

Pros:
* Clearer what's going on
* Would give syntax for changing the monomial basis for the SOS proofs (``Poly(monomials=Z)``), which is a feature worth having when working with domains in this way anyway.

**Comment by mlubin** [comment]
> Though this would probably be an indication that the two packages aren't compatible to begin with.

Not being compatible and one being unusable whenever the other is loaded are very different.

**Comment by blegat** [comment]
@joehuchette how would you do it with the dispatch ? It is tricky since the dispatch happens at "run time"/"after parse time" so Julia has tried to execute the expression given in domain at that time.

**Comment by joehuchette** [comment]
True, we would have to hold onto the raw expressions until runtime, which isn't great.

**Comment by joehuchette** [comment]
OK, I guess I can live with ``@set`` for the meanwhile, but I'd like to try to think up a better solution at the meet-up.

**Pull request merged**